scripts/get-all-images.py

```diff
@@ -0,0 +1,223 @@
+        return True
+
+    return False
+
+
+def bundle_app_contains_gh_metadata(app: dict) -> bool:
+    """
+    Given an application in a bundle check if it contains github metadata keys
+    to be able to be parsed properly.
+
+    Args:
+        app(dict): app metadata from a bundle.yaml in dictionary form
+
+    Returns:
+        True if app has GH metadata, else False
+    """
+    if is_dependency_app(app):
+        return True
+
+    if GH_REPO_KEY in app and GH_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def clone_repo(repo_name: str, branch: str) -> git.PathLike:
+    """
+    Clones locally a repo and returns the folder created.
+
+    Args:
+        repo_name(str): name of the repo to clone
+        branch(str): branch to checkout to, once cloned the repo
+
+    Returns:
+        The path of the downloaded repository
+    """
+    repo_url = f"https://github.com/canonical/{repo_name}.git"
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo.working_dir
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    try:
+        process = subprocess.run(["bash", "tools/get-images.sh"],
+                                 cwd=repo_dir, capture_output=True,
+                                 text=True, check=True)
+    except subprocess.CalledProcessError as exc:
+        logging.error("Script '%s' for charm '%s' failed: %s",
+                      GET_IMAGES_SH, app["charm"], exc.stderr)
+        raise exc
+    finally:
+        # cleanup
+        shutil.rmtree(repo_dir)
+
+    images = process.stdout.split("\n")
+
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
```

Comment:
Nit. Maybe there is an easier way of doing this with yaml? Another thing we could do is use python `sh` and call `yq` to extract the values. Not a big blocker.

Reply:
I do not agree to do it with `yq`, it will bring just more complexity to the code and one more dependency, which this script could not controller. Let say I want to run in locally, but I do not have `yq` installed, or I have different version.

Reply:
Sounds good, then